// @vidstack/react/dist/prod/chunks/vidstack-412a7593.js
"use client"

import { p as peek, D as DOMEvent, l as listenEvent, e as effect, c as camelToKebabCase, i as isUndefined, a as isString, b as isFunction } from './vidstack-c3208662.js';
import { Q as QualitySymbol, T as TextTrack, a as TextTrackSymbol, L as ListSymbol, I as IS_CHROME, c as coerceToError, l as loadScript, p as preconnect, i as isHLSSupported } from './vidstack-16cb0cf6.js';
import { VideoProvider } from './vidstack-335f14e5.js';
import { R as RAFLoop } from './vidstack-5d47fadb.js';
import 'react';
import 'react-dom';

const toDOMEventType = (type) => camelToKebabCase(type);
class HLSController {
  constructor(_video) {
    this.Da = _video;
    this.ua = null;
    this.zi = null;
    this.ig = {};
    this.H = /* @__PURE__ */ new Set();
  }
  get instance() {
    return this.ua;
  }
  setup(ctor, context) {
    this.yi = context;
    const isLive = peek(context.$state.streamType).includes("live"), isLiveLowLatency = peek(context.$state.streamType).includes("ll-");
    this.ua = new ctor({
      lowLatencyMode: isLiveLowLatency,
      backBufferLength: isLiveLowLatency ? 4 : isLive ? 8 : void 0,
      renderTextTracksNatively: false,
      ...this.ig
    });
    const dispatcher = this.Ai.bind(this);
    for (const event of Object.values(ctor.Events))
      this.ua.on(event, dispatcher);
    this.ua.on(ctor.Events.ERROR, this.od.bind(this));
    for (const callback of this.H)
      callback(this.ua);
    context.player.dispatch(new DOMEvent("hls-instance", { detail: this.ua }));
    this.ua.attachMedia(this.Da);
    this.ua.on(ctor.Events.AUDIO_TRACK_SWITCHED, this.Bi.bind(this));
    this.ua.on(ctor.Events.LEVEL_SWITCHED, this.Ci.bind(this));
    this.ua.on(ctor.Events.LEVEL_LOADED, this.Di.bind(this));
    this.ua.on(ctor.Events.NON_NATIVE_TEXT_TRACKS_FOUND, this.Ei.bind(this));
    this.ua.on(ctor.Events.CUES_PARSED, this.Fi.bind(this));
    context.qualities[QualitySymbol.pb] = this.Gi.bind(this);
    listenEvent(context.qualities, "change", this.Fh.bind(this));
    listenEvent(context.audioTracks, "change", this.Hi.bind(this));
    this.zi = effect(this.Ii.bind(this));
  }
  Ii() {
    if (!this.yi.$state.live())
      return;
    const raf = new RAFLoop(this.Ji.bind(this));
    raf.pc();
    return raf.qc.bind(raf);
  }
  Ji() {
    this.yi.$state.liveSyncPosition.set(this.ua?.liveSyncPosition ?? Infinity);
  }
  Ai(eventType, detail) {
    this.yi.player?.dispatch(new DOMEvent(toDOMEventType(eventType), { detail }));
  }
  Ei(eventType, data) {
    const event = new DOMEvent(eventType, { detail: data });
    let currentTrack = -1;
    for (let i = 0; i < data.tracks.length; i++) {
      const nonNativeTrack = data.tracks[i], init = nonNativeTrack.subtitleTrack ?? nonNativeTrack.closedCaptions, track = new TextTrack({
        id: `hls-${nonNativeTrack.kind}${i}`,
        src: init?.url,
        label: nonNativeTrack.label,
        language: init?.lang,
        kind: nonNativeTrack.kind
      });
      track[TextTrackSymbol.ya] = 2;
      track[TextTrackSymbol.Ta] = () => {
        if (track.mode === "showing") {
          this.ua.subtitleTrack = i;
          currentTrack = i;
        } else if (currentTrack === i) {
          this.ua.subtitleTrack = -1;
          currentTrack = -1;
        }
      };
      if (nonNativeTrack.default)
        track.setMode("showing", event);
      this.yi.textTracks.add(track, event);
    }
  }
  Fi(eventType, data) {
    const track = this.yi.textTracks.getById(`hls-${data.track}`);
    if (!track)
      return;
    const event = new DOMEvent(eventType, { detail: data });
    for (const cue of data.cues) {
      cue.positionAlign = "auto";
      track.addCue(cue, event);
    }
  }
  Bi(eventType, data) {
    const track = this.yi.audioTracks[data.id];
    if (track) {
      this.yi.audioTracks[ListSymbol.$](
        track,
        true,
        new DOMEvent(eventType, { detail: data })
      );
    }
  }
  Ci(eventType, data) {
    const quality = this.yi.qualities[data.level];
    if (quality) {
      this.yi.qualities[ListSymbol.$](
        quality,
        true,
        new DOMEvent(eventType, { detail: data })
      );
    }
  }
  Di(eventType, data) {
    if (this.yi.$state.canPlay())
      return;
    const { type, live, totalduration: duration, targetduration } = data.details;
    const event = new DOMEvent(eventType, { detail: data });
    this.yi.delegate.Nb("stream-type-change", {
      detail: live ? type === "EVENT" && Number.isFinite(duration) && targetduration >= 10 ? "live:dvr" : "live" : "on-demand",
      trigger: event
    });
    this.yi.delegate.Nb("duration-change", { detail: duration, trigger: event });
    const media = this.ua.media;
    if (this.ua.currentLevel === -1) {
      this.yi.qualities[QualitySymbol.ob](true, event);
    }
    for (const track of this.ua.audioTracks) {
      this.yi.audioTracks[ListSymbol.h](
        {
          id: track.id + "",
          label: track.name,
          language: track.lang || "",
          kind: "main"
        },
        event
      );
    }
    for (const level of this.ua.levels) {
      this.yi.qualities[ListSymbol.h](
        {
          width: level.width,
          height: level.height,
          codec: level.codecSet,
          bitrate: level.bitrate
        },
        event
      );
    }
    media.dispatchEvent(new DOMEvent("canplay", { trigger: event }));
  }
  od(eventType, data) {
    if (data.fatal) {
      switch (data.type) {
        case "networkError":
          this.ua?.startLoad();
          break;
        case "mediaError":
          this.ua?.recoverMediaError();
          break;
        default:
          this.ua?.destroy();
          this.ua = null;
          break;
      }
    }
  }
  Gi() {
    if (this.ua)
      this.ua.currentLevel = -1;
  }
  Fh() {
    const { qualities } = this.yi;
    if (!this.ua || qualities.auto)
      return;
    this.ua[qualities.switch + "Level"] = qualities.selectedIndex;
    if (IS_CHROME)
      this.Da.currentTime = this.Da.currentTime;
  }
  Hi() {
    const { audioTracks } = this.yi;
    if (this.ua && this.ua.audioTrack !== audioTracks.selectedIndex) {
      this.ua.audioTrack = audioTracks.selectedIndex;
    }
  }
  N() {
    if (this.yi)
      this.yi.qualities[QualitySymbol.pb] = void 0;
    this.ua?.destroy();
    this.ua = null;
    this.zi?.();
    this.zi = null;
  }
}

class HLSLibLoader {
  constructor(_lib, _ctx, _callback) {
    this.zf = _lib;
    this.yi = _ctx;
    this.Bg = _callback;
    this.Ki();
  }
  async Ki() {
    const callbacks = {
      onLoadStart: this.Kf.bind(this),
      onLoaded: this.Vf.bind(this),
      onLoadError: this.Li.bind(this)
    };
    let ctor = await loadHLSScript(this.zf, callbacks);
    if (isUndefined(ctor) && !isString(this.zf))
      ctor = await importHLS(this.zf, callbacks);
    if (!ctor)
      return null;
    if (!ctor.isSupported()) {
      const message = "[vidstack]: `hls.js` is not supported in this environment";
      this.yi.player.dispatch(new DOMEvent("hls-unsupported"));
      this.yi.delegate.Nb("error", { detail: { message, code: 4 } });
      return null;
    }
    return ctor;
  }
  Kf() {
    this.yi.player.dispatch(new DOMEvent("hls-lib-load-start"));
  }
  Vf(ctor) {
    this.yi.player.dispatch(
      new DOMEvent("hls-lib-loaded", {
        detail: ctor
      })
    );
    this.Bg(ctor);
  }
  Li(e) {
    const error = coerceToError(e);
    this.yi.player.dispatch(
      new DOMEvent("hls-lib-load-error", {
        detail: error
      })
    );
    this.yi.delegate.Nb("error", {
      detail: { message: error.message, code: 4 }
    });
  }
}
async function importHLS(loader, callbacks = {}) {
  if (isUndefined(loader))
    return void 0;
  callbacks.onLoadStart?.();
  if (loader.prototype && loader.prototype !== Function) {
    callbacks.onLoaded?.(loader);
    return loader;
  }
  try {
    const ctor = (await loader())?.default;
    if (ctor && !!ctor.isSupported) {
      callbacks.onLoaded?.(ctor);
    } else {
      throw Error(
        false ? "[vidstack] failed importing `hls.js`. Dynamic import returned invalid constructor." : ""
      );
    }
    return ctor;
  } catch (err) {
    callbacks.onLoadError?.(err);
  }
  return void 0;
}
async function loadHLSScript(src, callbacks = {}) {
  if (!isString(src))
    return void 0;
  callbacks.onLoadStart?.();
  try {
    await loadScript(src);
    if (!isFunction(window.Hls)) {
      throw Error(
        false ? "[vidstack] failed loading `hls.js`. Could not find a valid `Hls` constructor on window" : ""
      );
    }
    const ctor = window.Hls;
    callbacks.onLoaded?.(ctor);
    return ctor;
  } catch (err) {
    callbacks.onLoadError?.(err);
  }
  return void 0;
}

const JS_DELIVR_CDN = "https://cdn.jsdelivr.net";
class HLSProvider extends VideoProvider {
  constructor() {
    super(...arguments);
    this.$$PROVIDER_TYPE = "HLS";
    this.hg = null;
    this.af = new HLSController(this.video);
    this.gg = `${JS_DELIVR_CDN}/npm/hls.js@^1.0.0/dist/hls${".min.js"}`;
  }
  /**
   * The `hls.js` constructor.
   */
  get ctor() {
    return this.hg;
  }
  /**
   * The current `hls.js` instance.
   */
  get instance() {
    return this.af.instance;
  }
  get type() {
    return "hls";
  }
  get canLiveSync() {
    return true;
  }
  /**
   * The `hls.js` configuration object.
   *
   * @see {@link https://github.com/video-dev/hls.js/blob/master/docs/API.md#fine-tuning}
   */
  get config() {
    return this.af.ig;
  }
  set config(config) {
    this.af.ig = config;
  }
  /**
   * The `hls.js` constructor (supports dynamic imports) or a URL of where it can be found.
   *
   * @defaultValue `https://cdn.jsdelivr.net/npm/hls.js@^1.0.0/dist/hls.min.js`
   */
  get library() {
    return this.gg;
  }
  set library(library) {
    this.gg = library;
  }
  preconnect() {
    if (!isString(this.gg))
      return;
    preconnect(this.gg);
  }
  setup(context) {
    super.setup(context);
    new HLSLibLoader(this.gg, context, (ctor) => {
      this.hg = ctor;
      this.af.setup(ctor, context);
      context.delegate.Nb("provider-setup", { detail: this });
      const src = peek(context.$state.source);
      if (src)
        this.loadSource(src);
    });
  }
  async loadSource(src, preload) {
    if (!isString(src.src))
      return;
    this.Ha.preload = preload || "";
    this.af.instance?.loadSource(src.src);
    this.jg = src;
  }
  /**
   * The given callback is invoked when a new `hls.js` instance is created and right before it's
   * attached to media.
   */
  onInstance(callback) {
    const instance = this.af.instance;
    if (instance)
      callback(instance);
    this.af.H.add(callback);
    return () => this.af.H.delete(callback);
  }
  destroy() {
    this.af.N();
  }
}
HLSProvider.supported = isHLSSupported();

export { HLSProvider };
