// @vidstack/react/dist/prod/chunks/vidstack-335f14e5.js
"use client"

import { o as onDispose, l as listenEvent, D as DOMEvent, s as scoped } from './vidstack-c3208662.js';
import { T as TextTrack, a as TextTrackSymbol, b as canUsePictureInPicture, d as canUseVideoPresentation, e as canPlayHLSNatively } from './vidstack-16cb0cf6.js';
import { H as HTMLMediaProvider } from './vidstack-5d47fadb.js';
import 'react';
import 'react-dom';

class NativeHLSTextTracks {
  constructor(_video, _context) {
    this.Da = _video;
    this.R = _context;
    _video.textTracks.onaddtrack = this.Ja.bind(this);
    onDispose(this.si.bind(this));
  }
  Ja(event) {
    const nativeTrack = event.track;
    if (!nativeTrack || findTextTrackElement(this.Da, nativeTrack))
      return;
    const track = new TextTrack({
      id: nativeTrack.id,
      kind: nativeTrack.kind,
      label: nativeTrack.label,
      language: nativeTrack.language,
      type: "vtt"
    });
    track[TextTrackSymbol._a] = { track: nativeTrack };
    track[TextTrackSymbol.ya] = 2;
    track[TextTrackSymbol.Na] = true;
    let lastIndex = 0;
    const onCueChange = (event2) => {
      if (!nativeTrack.cues)
        return;
      for (let i = lastIndex; i < nativeTrack.cues.length; i++) {
        track.addCue(nativeTrack.cues[i], event2);
        lastIndex++;
      }
    };
    onCueChange(event);
    nativeTrack.oncuechange = onCueChange;
    this.R.textTracks.add(track, event);
    track.setMode(nativeTrack.mode, event);
  }
  si() {
    this.Da.textTracks.onaddtrack = null;
    for (const track of this.R.textTracks) {
      const nativeTrack = track[TextTrackSymbol._a]?.track;
      if (nativeTrack?.oncuechange)
        nativeTrack.oncuechange = null;
    }
  }
}
function findTextTrackElement(video, track) {
  return Array.from(video.children).find((el) => el.track === track);
}

class VideoPictureInPicture {
  constructor(_video, _media) {
    this.Da = _video;
    this.Ha = _media;
    this.oa = (active, event) => {
      this.Ha.delegate.Nb("picture-in-picture-change", {
        detail: active,
        trigger: event
      });
    };
    listenEvent(this.Da, "enterpictureinpicture", this.ti.bind(this));
    listenEvent(this.Da, "leavepictureinpicture", this.ui.bind(this));
  }
  get active() {
    return document.pictureInPictureElement === this.Da;
  }
  get supported() {
    return canUsePictureInPicture(this.Da);
  }
  async enter() {
    return this.Da.requestPictureInPicture();
  }
  exit() {
    return document.exitPictureInPicture();
  }
  ti(event) {
    this.oa(true, event);
  }
  ui(event) {
    this.oa(false, event);
  }
}

class VideoPresentation {
  constructor(_video, _media) {
    this.Da = _video;
    this.Ha = _media;
    this.Xa = "inline";
    listenEvent(this.Da, "webkitpresentationmodechanged", this.Ta.bind(this));
  }
  get xi() {
    return canUseVideoPresentation(this.Da);
  }
  async wi(mode) {
    if (this.Xa === mode)
      return;
    this.Da.webkitSetPresentationMode(mode);
  }
  Ta() {
    const prevMode = this.Xa;
    this.Xa = this.Da.webkitPresentationMode;
    this.Ha.player?.dispatch(
      new DOMEvent("video-presentation-change", {
        detail: this.Xa,
        trigger: event
      })
    );
    ["fullscreen", "picture-in-picture"].forEach((type) => {
      if (this.Xa === type || prevMode === type) {
        this.Ha.delegate.Nb(`${type}-change`, {
          detail: this.Xa === type,
          trigger: event
        });
      }
    });
  }
}
class FullscreenPresentationAdapter {
  constructor(_presentation) {
    this.vi = _presentation;
  }
  get active() {
    return this.vi.Xa === "fullscreen";
  }
  get supported() {
    return this.vi.xi;
  }
  async enter() {
    this.vi.wi("fullscreen");
  }
  async exit() {
    this.vi.wi("inline");
  }
}
class PIPPresentationAdapter {
  constructor(_presentation) {
    this.vi = _presentation;
  }
  get active() {
    return this.vi.Xa === "picture-in-picture";
  }
  get supported() {
    return this.vi.xi;
  }
  async enter() {
    this.vi.wi("picture-in-picture");
  }
  async exit() {
    this.vi.wi("inline");
  }
}

class VideoProvider extends HTMLMediaProvider {
  constructor(video, context) {
    super(video);
    this.$$PROVIDER_TYPE = "VIDEO";
    scoped(() => {
      if (canUseVideoPresentation(video)) {
        const presentation = new VideoPresentation(video, context);
        this.fullscreen = new FullscreenPresentationAdapter(presentation);
        this.pictureInPicture = new PIPPresentationAdapter(presentation);
      } else if (canUsePictureInPicture(video)) {
        this.pictureInPicture = new VideoPictureInPicture(video, context);
      }
    }, this.scope);
  }
  get type() {
    return "video";
  }
  setup(context) {
    super.setup(context);
    if (canPlayHLSNatively(this.video)) {
      new NativeHLSTextTracks(this.video, context);
    }
    context.textRenderers.Ma(this.video);
    onDispose(() => {
      context.textRenderers.Ma(null);
    });
    if (this.type === "video")
      context.delegate.Nb("provider-setup", { detail: this });
  }
  /**
   * The native HTML `<video>` element.
   *
   * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/HTMLVideoElement}
   */
  get video() {
    return this.Ha;
  }
}

export { VideoProvider };
