// @vidstack/react/dist/prod/chunks/vidstack-5d47fadb.js
"use client"

import { i as isUndefined, d as isNumber, u as useDisposalBin, e as effect, o as onDispose, l as listenEvent, p as peek, D as DOMEvent, f as isNil, g as createScope, h as setAttribute, a as isString } from './vidstack-c3208662.js';
import { f as IS_SAFARI, g as isHLSSrc, h as getNumberOfDecimalPlaces, L as ListSymbol, j as isMediaStream } from './vidstack-16cb0cf6.js';

class RAFLoop {
  constructor(_callback) {
    this.Bg = _callback;
  }
  pc() {
    if (!isUndefined(this.zc))
      return;
    this.xj();
  }
  qc() {
    if (isNumber(this.zc))
      window.cancelAnimationFrame(this.zc);
    this.zc = void 0;
  }
  xj() {
    this.zc = window.requestAnimationFrame(() => {
      if (isUndefined(this.zc))
        return;
      this.Bg();
      this.xj();
    });
  }
}

class HTMLMediaEvents {
  constructor(_provider, _ctx) {
    this.Rb = _provider;
    this.yi = _ctx;
    this.qi = useDisposalBin();
    this.Pi = false;
    this.Ri = false;
    this.Si = false;
    this.Qi = new RAFLoop(this.Vi.bind(this));
    this.Wi = void 0;
    this.pj = void 0;
    this.Xi();
    effect(this.Yi.bind(this));
    onDispose(this.si.bind(this));
  }
  get Ha() {
    return this.Rb.media;
  }
  get Xc() {
    return this.yi.delegate;
  }
  si() {
    this.Ri = false;
    this.Si = false;
    this.Qi.qc();
    this.qi.empty();
  }
  /**
   * The `timeupdate` event fires surprisingly infrequently during playback, meaning your progress
   * bar (or whatever else is synced to the currentTime) moves in a choppy fashion. This helps
   * resolve that by retrieving time updates in a request animation frame loop.
   */
  Vi() {
    const newTime = this.Rb.currentTime;
    if (this.yi.$state.currentTime() !== newTime)
      this.Oi(newTime);
  }
  Xi() {
    this.Ni("loadstart", this.Kf);
    this.Ni("abort", this.Ui);
    this.Ni("emptied", this.Zi);
    this.Ni("error", this.od);
  }
  _i() {
    if (this.Ri)
      return;
    this.qi.add(
      this.Ni("loadeddata", this.$i),
      this.Ni("loadedmetadata", this.aj),
      this.Ni("canplay", this.hd),
      this.Ni("canplaythrough", this.bj),
      this.Ni("durationchange", this.cj),
      this.Ni("play", this.qa),
      this.Ni("progress", this.dj),
      this.Ni("stalled", this.ej),
      this.Ni("suspend", this.fj)
    );
    this.Ri = true;
  }
  gj() {
    if (this.Si)
      return;
    this.qi.add(
      this.Ni("pause", this.ra),
      this.Ni("playing", this.hj),
      this.Ni("ratechange", this.ij),
      this.Ni("seeked", this.jj),
      this.Ni("seeking", this.kj),
      this.Ni("ended", this.lj),
      this.Ni("volumechange", this.vd),
      this.Ni("waiting", this.mj)
    );
    this.Si = true;
  }
  Ni(eventType, handler) {
    return listenEvent(
      this.Ha,
      eventType,
      handler.bind(this)
    );
  }
  qj(event2) {
    return;
  }
  Oi(time, trigger) {
    this.Xc.Nb("time-update", {
      // Avoid errors where `currentTime` can have higher precision.
      detail: {
        currentTime: Math.min(time, this.yi.$state.seekableEnd()),
        played: this.Ha.played
      },
      trigger
    });
  }
  Kf(event2) {
    if (this.Ha.networkState === 3) {
      this.Ui(event2);
      return;
    }
    this._i();
    this.Xc.Nb("load-start", { trigger: event2 });
  }
  Ui(event2) {
    this.Xc.Nb("abort", { trigger: event2 });
  }
  Zi() {
    this.Xc.Nb("emptied", { trigger: event });
  }
  $i(event2) {
    this.Xc.Nb("loaded-data", { trigger: event2 });
  }
  aj(event2) {
    this.gj();
    this.Xc.Nb("volume-change", {
      detail: {
        volume: this.Ha.volume,
        muted: this.Ha.muted
      }
    });
    this.Xc.Nb("loaded-metadata", { trigger: event2 });
    if (IS_SAFARI && isHLSSrc(this.yi.$state.source())) {
      this.Xc.Dg(this.Ti(), event2);
    }
  }
  Ti() {
    return {
      provider: peek(this.yi.$provider),
      duration: this.Ha.duration,
      buffered: this.Ha.buffered,
      seekable: this.Ha.seekable
    };
  }
  qa(event2) {
    if (!this.yi.$state.canPlay)
      return;
    this.Xc.Nb("play", { trigger: event2 });
  }
  ra(event2) {
    if (this.Ha.readyState === 1 && !this.Pi)
      return;
    this.Pi = false;
    this.Qi.qc();
    this.Xc.Nb("pause", { trigger: event2 });
  }
  hd(event2) {
    this.Xc.Dg(this.Ti(), event2);
  }
  bj(event2) {
    if (this.yi.$state.started())
      return;
    this.Xc.Nb("can-play-through", {
      trigger: event2,
      detail: this.Ti()
    });
  }
  hj(event2) {
    this.Pi = false;
    this.Xc.Nb("playing", { trigger: event2 });
    this.Qi.pc();
  }
  ej(event2) {
    this.Xc.Nb("stalled", { trigger: event2 });
    if (this.Ha.readyState < 3) {
      this.Pi = true;
      this.Xc.Nb("waiting", { trigger: event2 });
    }
  }
  mj(event2) {
    if (this.Ha.readyState < 3) {
      this.Pi = true;
      this.Xc.Nb("waiting", { trigger: event2 });
    }
  }
  lj(event2) {
    this.Qi.qc();
    this.Oi(this.Ha.duration, event2);
    this.Xc.Nb("end", { trigger: event2 });
    if (this.yi.$state.loop()) {
      this.nj();
    } else {
      this.Xc.Nb("ended", { trigger: event2 });
    }
  }
  Yi() {
    if (this.yi.$state.paused()) {
      listenEvent(this.Ha, "timeupdate", this.oj.bind(this));
    }
  }
  oj(event2) {
    this.Oi(this.Ha.currentTime, event2);
  }
  cj(event2) {
    if (this.yi.$state.ended()) {
      this.Oi(this.Ha.duration, event2);
    }
    this.Xc.Nb("duration-change", {
      detail: this.Ha.duration,
      trigger: event2
    });
  }
  vd(event2) {
    this.Xc.Nb("volume-change", {
      detail: {
        volume: this.Ha.volume,
        muted: this.Ha.muted
      },
      trigger: event2
    });
  }
  jj(event2) {
    this.Oi(this.Ha.currentTime, event2);
    this.Xc.Nb("seeked", {
      detail: this.Ha.currentTime,
      trigger: event2
    });
    if (Math.trunc(this.Ha.currentTime) === Math.trunc(this.Ha.duration) && getNumberOfDecimalPlaces(this.Ha.duration) > getNumberOfDecimalPlaces(this.Ha.currentTime)) {
      this.Oi(this.Ha.duration, event2);
      if (!this.Ha.ended) {
        this.yi.player.dispatch(
          new DOMEvent("media-play-request", {
            trigger: event2
          })
        );
      }
    }
  }
  kj(event2) {
    this.Xc.Nb("seeking", {
      detail: this.Ha.currentTime,
      trigger: event2
    });
  }
  dj(event2) {
    this.Xc.Nb("progress", {
      detail: {
        buffered: this.Ha.buffered,
        seekable: this.Ha.seekable
      },
      trigger: event2
    });
  }
  nj() {
    const hasCustomControls = isNil(this.Ha.controls);
    if (hasCustomControls)
      this.Ha.controls = false;
    this.yi.player.dispatch(new DOMEvent("media-loop-request"));
  }
  fj(event2) {
    this.Xc.Nb("suspend", { trigger: event2 });
  }
  ij(event2) {
    this.Xc.Nb("rate-change", {
      detail: this.Ha.playbackRate,
      trigger: event2
    });
  }
  od(event2) {
    const error = this.Ha.error;
    if (!error)
      return;
    this.Xc.Nb("error", {
      detail: {
        message: error.message,
        code: error.code,
        mediaError: error
      },
      trigger: event2
    });
  }
}

class NativeAudioTracks {
  constructor(_provider, _context) {
    this.Rb = _provider;
    this.R = _context;
    this.rj.onaddtrack = this.tj.bind(this);
    this.rj.onremovetrack = this.uj.bind(this);
    this.rj.onchange = this.vj.bind(this);
    listenEvent(this.R.audioTracks, "change", this.wj.bind(this));
  }
  get rj() {
    return this.Rb.media.audioTracks;
  }
  tj(event) {
    const _track = event.track;
    if (_track.label === "")
      return;
    const audioTrack = {
      id: _track.id + "",
      label: _track.label,
      language: _track.language,
      kind: _track.kind,
      selected: false
    };
    this.R.audioTracks[ListSymbol.h](audioTrack, event);
    if (_track.enabled)
      audioTrack.selected = true;
  }
  uj(event) {
    const track = this.R.audioTracks.getById(event.track.id);
    if (track)
      this.R.audioTracks[ListSymbol.f](track, event);
  }
  vj(event) {
    let enabledTrack = this.sj();
    if (!enabledTrack)
      return;
    const track = this.R.audioTracks.getById(enabledTrack.id);
    if (track)
      this.R.audioTracks[ListSymbol.$](track, true, event);
  }
  sj() {
    return Array.from(this.rj).find((track) => track.enabled);
  }
  wj(event) {
    const { current } = event.detail;
    if (!current)
      return;
    const track = this.rj.getTrackById(current.id);
    if (track) {
      const prev = this.sj();
      if (prev)
        prev.enabled = false;
      track.enabled = true;
    }
  }
}

class HTMLMediaProvider {
  constructor(_media) {
    this.Ha = _media;
    this.scope = createScope();
    this.jg = null;
  }
  setup(context) {
    new HTMLMediaEvents(this, context);
    if ("audioTracks" in this.media)
      new NativeAudioTracks(this, context);
    onDispose(() => {
      this.Ha.setAttribute("src", "");
      this.Ha.load();
    });
  }
  get type() {
    return "";
  }
  get media() {
    return this.Ha;
  }
  get currentSrc() {
    return this.jg;
  }
  get paused() {
    return this.Ha.paused;
  }
  get muted() {
    return this.Ha.muted;
  }
  set muted(muted) {
    this.Ha.muted = muted;
  }
  get volume() {
    return this.Ha.volume;
  }
  set volume(volume) {
    this.Ha.volume = volume;
  }
  get currentTime() {
    return this.Ha.currentTime;
  }
  set currentTime(time) {
    this.Ha.currentTime = time;
  }
  get playsinline() {
    return this.Ha.hasAttribute("playsinline");
  }
  set playsinline(playsinline) {
    setAttribute(this.Ha, "playsinline", playsinline);
  }
  get playbackRate() {
    return this.Ha.playbackRate;
  }
  set playbackRate(rate) {
    this.Ha.playbackRate = rate;
  }
  async play() {
    return this.Ha.play();
  }
  async pause() {
    return this.Ha.pause();
  }
  async loadSource({ src, type }, preload) {
    this.Ha.preload = preload || "";
    if (isMediaStream(src)) {
      this.Ha.srcObject = src;
    } else {
      this.Ha.srcObject = null;
      this.Ha.src = isString(src) ? src : window.URL.createObjectURL(src);
    }
    this.Ha.load();
    this.jg = { src, type };
  }
}

export { HTMLMediaProvider as H, RAFLoop as R };
